src/operator/nn/im2col-inl.h

```diff
@@ -0,0 +1,247 @@
+
+    for (index_t n = 0; n < num; ++n) {
+      col2im(s, col_grad[n].dptr_, im_shape, col_buffer_shape,
+             param.kernel, param.pad, param.stride, param.dilate,
+             im_grad[n].dptr_, req[0]);
+    }
+  });
+}
+
+struct Col2imParam : public dmlc::Parameter<Col2imParam> {
+  mxnet::TShape output_size;
+  mxnet::TShape kernel;
+  mxnet::TShape stride;
+  mxnet::TShape dilate;
+  mxnet::TShape pad;
+  DMLC_DECLARE_PARAMETER(Col2imParam) {
+    DMLC_DECLARE_FIELD(output_size)
+    .describe("The spatial dimension of image array: (w,), (h, w) or (d, h, w).");
+    DMLC_DECLARE_FIELD(kernel).describe("Sliding kernel size: (w,), (h, w) or (d, h, w).");
+    DMLC_DECLARE_FIELD(stride).set_default(mxnet::TShape(0, 0))
+    .describe("The stride between adjacent sliding blocks in spatial dimension: "
+              "(w,), (h, w) or (d, h, w). Defaults to 1 for each dimension.");
+    DMLC_DECLARE_FIELD(dilate).set_default(mxnet::TShape(0, 0))
+    .describe("The spacing between adjacent kernel points: (w,), (h, w) or (d, h, w). "
+              "Defaults to 1 for each dimension.");
+    DMLC_DECLARE_FIELD(pad).set_default(mxnet::TShape(0, 0))
+    .describe("The zero-value padding size on both sides of spatial dimension: "
+              "(w,), (h, w) or (d, h, w). Defaults to no padding.");
+  }
+
+  index_t DilatedKernelSize(int dim) const {
+    return 1 + (kernel[dim] - 1) * dilate[dim];
+  }
+};  // struct Col2imParam
+
+template<typename xpu>
+void Col2imCompute(const nnvm::NodeAttrs& attrs,
+                   const OpContext& ctx,
+                   const std::vector<TBlob>& inputs,
+                   const std::vector<OpReqType>& req,
+                   const std::vector<TBlob>& outputs) {
+  using namespace mshadow;
+  const Col2imParam& param = nnvm::get<Col2imParam>(attrs.parsed);
+  Stream<xpu> *s = ctx.get_stream<xpu>();
+  const mxnet::TShape im_shape = outputs[0].shape_;
+  const mxnet::TShape col_shape = inputs[0].shape_;
+  const index_t num = im_shape[0];
+
+  const int spatial_size = param.kernel.ndim();
+  mxnet::TShape col_buffer_shape(1 + spatial_size, 1);
+  col_buffer_shape[0] = col_shape[1];
+  for (int i = 0; i < spatial_size; ++i) {
+    const index_t pad_size = im_shape[i + 2] + 2 * param.pad[i];
+    const index_t output_size = (pad_size - param.DilatedKernelSize(i)) / param.stride[i] + 1;
+    col_buffer_shape[i + 1] = output_size;
+  }
+
+  MSHADOW_REAL_TYPE_SWITCH(outputs[0].type_flag_, DType, {
+    Tensor<xpu, 4, DType> im = outputs[0].get_with_shape<xpu, 4, DType>(
+        Shape4(im_shape[0], im_shape[1], im_shape[2], im_shape[3]), s);
+    Tensor<xpu, 3, DType> col = inputs[0].get_with_shape<xpu, 3, DType>(
+        Shape3(col_shape[0], col_shape[1], col_shape[2]), s);
+
+    for (index_t n = 0; n < num; ++n) {
+      col2im(s, col[n].dptr_, im_shape, col_buffer_shape,
+             param.kernel, param.pad, param.stride, param.dilate,
+             im[n].dptr_, kWriteTo);
```

Comment:
Replace kWriteTo with req[0] ?

Reply:
All fixed. Req types are checked in forward function, too.